File: Project Enhancements/Brea Software Design and Engineering Enhancement/Brea_SoftDesignEngineeringEnhancement/Source/src/ContactService.cpp
#include "ContactService.h"
#include <stdexcept>

void ContactService::addContact(const std::string& id, const std::string& firstName, 
                               const std::string& lastName, const std::string& phone, 
                               const std::string& address) {
    //validate ID
    if (id.empty() || id.length() > 10) {
        throw std::invalid_argument("Invalid ID");
    }

    //add contact to map
    contacts.emplace(id, Contact(id, firstName, lastName, phone, address));
}

//finds and removes a contact from the map via its id
void ContactService::removeContact(const std::string& id) {
    contacts.erase(id);
}

//finds and updates the first name of a contact via its id using the given parameters
void ContactService::updateContactFirstName(const std::string& id, const std::string& firstName) {
    contacts.at(id).setFirstName(firstName);
}

//finds and updates the last name of a contact via its id using the given parameters
void ContactService::updateContactLastName(const std::string& id, const std::string& lastName) {
    contacts.at(id).setLastName(lastName);
}

//finds and updates the phone number of a contact via its id using the given parameters
void ContactService::updateContactPhone(const std::string& id, const std::string& phone) {
    contacts.at(id).setPhone(phone);
}

//finds and updates the address of a contact via its id using the given parameters
void ContactService::updateContactAddress(const std::string& id, const std::string& address) {
    contacts.at(id).setAddress(address);
}

//finds and returns a contact via its id
Contact ContactService::getContact(const std::string& id) {
    return contacts.at(id);
}

File: Project Enhancements/Brea Software Design and Engineering Enhancement/BreaCS320Proj1/src/main/ContactService.java
package main;

import java.util.HashMap;

public class ContactService {
	//collection of contacts with iD as key for easy access
	private HashMap<String, Contact> service = new HashMap<String, Contact>();
	
	//add Contact to collection
	public void addContact(String iD, String firstName, String lastName, String phone, String address) {
		//check iD before setting it
		if (iD.length() > 10 || iD == null) {
			throw new IllegalArgumentException("Invalid ID");
		}
		
		//create contact object
		Contact entry = new Contact(iD, firstName, lastName, phone, address);
		
		//add to map
		service.put(iD, entry);
	}
	
	//remove from map based on iD
	public void removeContact(String iD) {
		service.remove(iD);
	}
	
	//set firstName based on iD
	public void updateContactFirstName(String iD, String firstName) {
		service.get(iD).setFirstName(firstName);
	}
	
	//set lastName based on iD
	public void updateContactLastName(String iD, String lastName) {
		service.get(iD).setLastName(lastName);
	}
	
	//set phone based on iD
	public void updateContactPhone(String iD,String phone) {
		service.get(iD).setPhone(phone);
	}
	
	//set address based on iD
	public void updateContactAddress(String iD, String address) {
		service.get(iD).setAddress(address);
	}
	
	//get contact based on iD
	public Contact getContact(String iD) {
		return service.get(iD);
	}
	
}
